unexpected build failure with `failOnError=false`

code:

```java
+package sandbox

[] > stress
  TRUE > a
  TRUE > a
  a > @
```

result [compilation.log](https://github.com/objectionary/eo/files/9302205/compilation.log):

![image](https://user-images.githubusercontent.com/33983579/183965519-7f2c70a8-68c1-488e-a7de-3e365095cacb.png)
![image](https://user-images.githubusercontent.com/33983579/183965659-450cc5b7-34d3-4479-96ba-8776bc5b34f8.png)

@yegor256, I tried, but got the same build failure result. Does it really have to be a critical error that stops the pipeline?

```java
[INFO] ---------------------< org.eolang:c2eo-transpiler >---------------------
[INFO] Building c2eo-transpiler 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ c2eo-transpiler ---
[INFO] Deleting D:\Projects\EO\sandox\target
[INFO]
[INFO] --- eo-maven-plugin:0.28.11:register (default) @ c2eo-transpiler ---
[INFO] Registered 1 EO sources from .\:\Projects\EO\sandox\eo to .\:\Projects\EO\sandox\target\eo\foreign.csv, included [**.eo], excluded []
[INFO]
[INFO] --- eo-maven-plugin:0.28.11:assemble (default) @ c2eo-transpiler ---
[INFO] 0 attributes loaded from 35 stream(s) in 94ms, 47 saved, 409 ignored: []
[INFO] Parsed 1 .eo sources to XMIRs
[INFO] Adding optimization task for D:\Projects\EO\sandox\target\01-parse\c2eo\test.xmir
[INFO] Running 1 optimizations in parallel
[WARNING] XSL transformation 'not-empty-atoms' took too long 391ms (over 100ms)
[WARNING] XSL transformation 'not-empty-atoms' took too long 469ms (over 100ms)
[ERROR] net.sf.saxon.style.XSLText@7b93446b: Critical error identified: The name "a" is already in use;
[ERROR] #fatalError(): Processing terminated by xsl:message at line 32 in file:///; SystemID: file:///; Line#: 32; Column#: 17
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.937 s
[INFO] Finished at: 2022-11-10T20:20:36+03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eolang:eo-maven-plugin:0.28.11:assemble (default) on project c2eo-transpiler: Execution default of goal org.eolang:eo-maven-plugin:0.28.11:assemble failed: Failed to transform by net.sf.saxon.jaxp.TransformerImpl: [Processing terminated by xsl:message at line 32 in file:///; SystemID: file:///; Line#: 32; Column#: 17] (Processing terminated by xsl:message at line 32 in file:///; SystemID: file:///; Line#: 32; Column#: 17): java.lang.IllegalArgumentException: Failed to transform by net.sf.saxon.jaxp.TransformerImpl: [Processing terminated by xsl:message at line 32 in file:///; SystemID: file:///; Line#: 32; Column#: 17] (Processing terminated by xsl:message at line 32 in file:///; SystemID: file:///; Line#: 32; Column#: 17) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java
@@ -46,12 +46,17 @@ import org.eolang.maven.optimization.OptCached;
 import org.eolang.maven.optimization.OptSpy;
 import org.eolang.maven.optimization.OptTrain;
 import org.eolang.maven.optimization.Optimization;
+import org.eolang.parser.ParsingTrain;
 
 /**
  * Optimize XML files.
  *
  * @todo #1336:30min Make a number of threads in `exec()` method configurable
  *  via mojo parameter `threads`. Default value should be set to 4.
+ * @todo #1024:30min On "trackOptimizationSteps=true", in "02-steps" folder
+ * there are creates full list of XSL-s, even if "critical" error found---
+ * it just creates the same files. Need to stop creating new files, when
+ * critical error found.
  * @since 0.1
  */
 @Mojo(
@@ -236,7 +241,12 @@ public final class OptimizeMojo extends SafeMojo {
                     .resolve(tojo.get(AssembleMojo.ATTR_HASH))
             );
         }
-        return new OptTrain(opt, "/org/eolang/parser/fail-on-critical.xsl");
+        if (this.failOnError) {
+            opt = new OptTrain(opt, "/org/eolang/parser/fail-on-critical.xsl");
+        } else {
+            opt = new OptTrain(opt, new ParsingTrain().empty());
+        }
+        return opt;
     }
 
     /**

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/OptimizeMojo.java
@@ -52,11 +52,7 @@ import org.eolang.parser.ParsingTrain;
  * Optimize XML files.
  *
  * @todo #1336:30min Make a number of threads in `exec()` method configurable
- *  via mojo parameter `threads`. Default value should be set to 4.
- * @todo #1024:30min On "trackOptimizationSteps=true", in "02-steps" folder
- * there are creates full list of XSL-s, even if "critical" error found---
- * it just creates the same files. Need to stop creating new files, when
- * critical error found.
+ *   via mojo parameter `threads`. Default value should be set to 4.
  * @since 0.1
  */
 @Mojo(

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptTrain.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptTrain.java
@@ -110,7 +110,7 @@ public final class OptTrain implements Optimization {
      * @param delegate Optimizations that have to be done before.
      * @param shifts To apply
      */
-    private OptTrain(
+    public OptTrain(
         final Optimization delegate,
         final Train<Shift> shifts
     ) {

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptTrain.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/optimization/OptTrain.java
@@ -93,7 +93,7 @@ public final class OptTrain implements Optimization {
     }
 
     /**
-     * Constructor for single shift optimization.
+     * Constructor for single {@link StClasspath} optimization.
      *
      * @param delegate Optimizations that have to be done before.
      * @param xls File from classpath.

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/OptimizeMojoTest.java
@@ -23,6 +23,7 @@
  */
 package org.eolang.maven;
 
+import com.jcabi.matchers.XhtmlMatchers;
 import com.jcabi.xml.XML;
 import com.jcabi.xml.XMLDocument;
 import com.jcabi.xml.XSLDocument;
@@ -345,6 +346,53 @@ final class OptimizeMojoTest {
 
     @Test
     void stopsOnCritical(@TempDir final Path temp) throws Exception {
+        final Path src = temp.resolve("foo/main.eo");
+        new Home(temp).save(
+            String.join(
+                "\n",
+                "+package f\n",
+                "[args] > main",
+                "  seq > @",
+                "    TRUE > x",
+                "    FALSE > x\n"
+            ),
+            src
+        );
+        final Path target = temp.resolve("target");
+        final Path foreign = temp.resolve("eo-foreign.csv");
+        Catalogs.INSTANCE.make(foreign)
+            .add("foo.main")
+            .set(AssembleMojo.ATTR_SCOPE, "compile")
+            .set(AssembleMojo.ATTR_EO, src.toString());
+        new Moja<>(ParseMojo.class)
+            .with("targetDir", target.toFile())
+            .with("foreign", foreign.toFile())
+            .with("foreignFormat", "csv")
+            .with("cache", temp.resolve("cache/parsed"))
+            .execute();
+        new Moja<>(OptimizeMojo.class)
+            .with("targetDir", target.toFile())
+            .with("foreign", foreign.toFile())
+            .with("foreignFormat", "csv")
+            .with("failOnError", false)
+            .with("trackOptimizationSteps", true)
+            .execute();
+        MatcherAssert.assertThat(
+            new XMLDocument(
+                target.resolve(
+                    String.format("%s/foo/main/02-duplicate-names.xml", OptimizeMojo.STEPS)
+                )
+            ),
+            XhtmlMatchers.hasXPaths(
+                "/program/sheets[count(sheet)=3]",
+                "/program/errors[count(error)=1]",
+                "/program/errors/error[@severity='critical']"
+            )
+        );
+    }
+
+    @Test
+    void failsOnCritical(@TempDir final Path temp) throws Exception {
         final Path src = temp.resolve("foo/main.eo");
         new Home(temp).save(
             String.join(

--- a/eo-parser/src/main/java/org/eolang/parser/ParsingTrain.java
+++ b/eo-parser/src/main/java/org/eolang/parser/ParsingTrain.java
@@ -40,6 +40,8 @@ import com.yegor256.xsline.TrLogged;
  * Train of XSL shifts.
  *
  * @since 0.1
+ * @todo #1024:30min Need to figure out, which errors need to be
+ * "critical"---same as "duplicate-names"
  */
 public final class ParsingTrain extends TrEnvelope {
 

--- a/eo-parser/src/main/java/org/eolang/parser/ParsingTrain.java
+++ b/eo-parser/src/main/java/org/eolang/parser/ParsingTrain.java
@@ -41,7 +41,8 @@ import com.yegor256.xsline.TrLogged;
  *
  * @since 0.1
  * @todo #1024:30min Need to figure out, which errors need to be
- * "critical"---same as "duplicate-names"
+ *   "critical", same as "duplicate-names" error. After that
+ *   move them to "critical-errors" directory.
  */
 public final class ParsingTrain extends TrEnvelope {
 
@@ -104,20 +105,18 @@ public final class ParsingTrain extends TrEnvelope {
                         ).back()
                     )
                 ),
-                shift -> new StLambda(
-                    shift::uid,
-                    (position, out) -> new StSequence(
-                        xml -> xml.nodes("//error[@severity='critical']").isEmpty(),
-                        new StAfter(
-                            shift,
-                            new StLambda(
-                                shift::uid,
-                                (pos, xml) -> ParsingTrain.EACH.with("step", pos)
-                                    .with("sheet", shift.uid())
-                                    .transform(xml)
-                            )
+                shift -> new StSequence(
+                    shift.uid(),
+                    xml -> xml.nodes("//error[@severity='critical']").isEmpty(),
+                    new StAfter(
+                        shift,
+                        new StLambda(
+                            shift::uid,
+                            (pos, xml) -> ParsingTrain.EACH.with("step", pos)
+                                .with("sheet", shift.uid())
+                                .transform(xml)
                         )
-                    ).apply(position, out)
+                    )
                 )
             )
         );

